reservation_units/querysets.py

```diff
@@ -0,0 +1,418 @@
+    a reservation must start at 00:00, 00:30, 01:00, 01:30 from the start of the time span.
+    """
+    interval = INTERVAL_TO_MINUTES[reservation_unit.reservation_start_interval]
+
+    if filter_time_start and selected_start_datetime.time() < filter_time_start:
+        selected_start_datetime = datetime.combine(
+            selected_start_datetime.date(), filter_time_start, tzinfo=DEFAULT_TIMEZONE
+        )
+
+    delta_minutes = (selected_start_datetime - time_span.start_datetime).total_seconds() / 60
+    delta_to_next_interval = timedelta(minutes=interval) * ceil(delta_minutes / interval)
+    return time_span.start_datetime + delta_to_next_interval
+
+
+class ReservationUnitInfo(TypedDict):
+    pk: int
+    space_ids: list[int]
+    resource_ids: list[int]
+
+
+class ReservationUnitQuerySet(SearchResultsQuerySet):
+    def scheduled_for_publishing(self):
+        now = datetime.now(tz=DEFAULT_TIMEZONE)
+        return self.filter(
+            Q(is_archived=False, is_draft=False)
+            & (
+                Q(publish_begins__isnull=False, publish_begins__gt=now)
+                | Q(publish_ends__isnull=False, publish_ends__lte=now)
+            )
+        )
+
+    def with_first_reservable_time(
+        self,
+        filter_date_start: date | None,
+        filter_date_end: date | None,
+        filter_time_start: time | None,
+        filter_time_end: time | None,
+        minimum_duration_minutes: int | float | Decimal | None,
+    ) -> Self:
+        """
+        Annotate the queryset with `first_reservable_datetime` and `is_closed` fields.
+
+        Date and Time filters are used to filter a range of dates and time when the reservation must be within.
+
+        This method works by first generating a list of closed time spans for each ReservationUnit and then one by one
+        checking the reservable time spans for a time span that is not overlapping with closed time spans and still
+        long enough to fulfill the minimum duration criteria.
+
+        The list of closed time spans are generated from:
+        - Given filter value
+        - ReservationUnit settings
+        - ReservationUnit's ApplicationRounds
+        - Reservations from ReservationUnit's TILAHIERARKIA TODO
+        After removing the closed time spans, the first reservable time span that is long enough to fit the minimum
+        duration is selected as the `first_reservable_datetime`.
+
+        Variables which affect the first reservable time span and cause the ReservationUnit to be "closed":
+        - Date filter range
+        - Time filter range
+        - ReservationUnit `reservation_begins`
+        - ReservationUnit `reservation_ends`
+        - ReservationUnit `publish_ends`
+        - ApplicationRound `reservation_period_begin`
+        - ApplicationRound `reservation_period_end`
+
+        Variables which only affect the first reservable time span and allow the ReservationUnit to be "open":
```

Comment:
Also clarify the condition here:

> ... first reservable time span, but allow the ReservationUnit to be "open" if there would otherwise be valid time spans on the filtered date range